stm32: Improve the CRC function to support fully accumulate

kevinong0108 opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->

Some STM32 series can support 32, 16, 8 and 7 bit hardware CRC but API only apply 32-bit.

Tested on DISCO_L476VG and NUCLEO_L476RG

Reopen: https://github.com/ARMmbed/mbed-os/pull/7951

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [x] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!-- 
    Optional
    Request additional reviewers with @username
-->
@jeromecoutant, @bcostm

kevinong0108 committed:
stm32: Improve the CRC function to support fully accumulate

Tested on DISCO_L476VG and NUCLEO_L476RG

kevinong0108 committed:
stm32: remove empty line

jeromecoutant commented:
Hi

I didn't review the updated code yet,

but tests results are quite good:

| target            | platform_name | test suite             | result | elapsed_time (sec) | copy_method |
|-------------------|---------------|------------------------|--------|--------------------|-------------|
| NUCLEO_F091RC-ARM | NUCLEO_F091RC | tests-mbed_drivers-crc | OK     | 17.99              | default     |
| NUCLEO_F091RC-ARM | NUCLEO_F091RC | tests-mbed_hal-crc     | OK     | 18.41              | default     |
| NUCLEO_F103RB-ARM | NUCLEO_F103RB | tests-mbed_drivers-crc | OK     | 18.56              | default     |
| NUCLEO_F207ZG-ARM | NUCLEO_F207ZG | tests-mbed_drivers-crc | OK     | 16.99              | default     |
| NUCLEO_F303ZE-ARM | NUCLEO_F303ZE | tests-mbed_drivers-crc | OK     | 18.36              | default     |
| NUCLEO_F303ZE-ARM | NUCLEO_F303ZE | tests-mbed_hal-crc     | OK     | 18.78              | default     |
| NUCLEO_F446RE-ARM | NUCLEO_F446RE | tests-mbed_drivers-crc | OK     | 16.93              | default     |
| NUCLEO_F767ZI-ARM | NUCLEO_F767ZI | tests-mbed_drivers-crc | OK     | 16.54              | default     |
| NUCLEO_F767ZI-ARM | NUCLEO_F767ZI | tests-mbed_hal-crc     | OK     | 17.16              | default     |
| NUCLEO_H743ZI-ARM | NUCLEO_H743ZI | tests-mbed_drivers-crc | OK     | 16.58              | default     |
| NUCLEO_H743ZI-ARM | NUCLEO_H743ZI | tests-mbed_hal-crc     | OK     | 17.33              | default     |
| NUCLEO_L073RZ-ARM | NUCLEO_L073RZ | tests-mbed_drivers-crc | OK     | 21.31              | default     |
| NUCLEO_L073RZ-ARM | NUCLEO_L073RZ | tests-mbed_hal-crc     | OK     | 20.64              | default     |
| NUCLEO_L152RE-ARM | NUCLEO_L152RE | tests-mbed_drivers-crc | OK     | 17.92              | default     |
| NUCLEO_L476RG-ARM | NUCLEO_L476RG | tests-mbed_drivers-crc | OK     | 17.52              | default     |
| NUCLEO_L476RG-ARM | NUCLEO_L476RG | tests-mbed_hal-crc     | OK     | 18.06              | default     |

jeromecoutant commented on targets/TARGET_STM/mbed_crc_api.c:
```diff
@@ -7,50 +7,103 @@
 
-    if (config->polynomial != POLY_32BIT_ANSI) {
+#if defined(TARGET_STM32F1) || defined(TARGET_STM32F2) || defined(TARGET_STM32F4) || defined(TARGET_STM32L1)
+    if (config->width != HAL_CRC_LENGTH_32B || config->polynomial != POLY_32BIT_ANSI ||
```
HAL_CRC_LENGTH_32B is not defined for STM32F1/F2/...,
so use direct value "32"

kevinong0108 replied:
OK, I will push new commit for this.

kevinong0108 replied:
Please verify this commit:
https://github.com/ARMmbed/mbed-os/pull/9347/commits/62a6870ddb516297ef073ed488e0be315fb2aef4

kevinong0108 replied:
I decide set F1/F2/F4/L1 to not supported, more detail please see ba25cd4

jeromecoutant reviewed (commented):
PR can be approved if you quickly want the update for the current targets supporting CRC
i.e. F0/F3/F7/L0/L4

But if you enable CRC at STM32_FAMILY, tests-mbed_hal-crc is still failing for the other families (F1/F2/F4/L1)

jeromecoutant commented on targets/TARGET_STM/mbed_crc_api.c:
```diff
@@ -7,50 +7,103 @@
+  | CRC initial Value       | Fixed to 0xFFFFFFFF   | Programmable  | Programmable  |
+  |                         |                       | on 32 bits    | on 8,16,32    |
+  +-------------------------+-----------------------+---------------+---------------+
+  | Handled data size in bit|                  32                   | 8,16,32       |
```
Seems that F0 supports 8-,16-, 32-bit data size?

kevinong0108 replied:
I also check F0 crc hal file from STM32Cube MCU Package before PR, it seems support right, but I was refer chapter 3 of [application note](https://www.st.com/content/ccc/resource/technical/document/application_note/39/89/da/89/9e/d7/49/b1/DM00068118.pdf/files/DM00068118.pdf/jcr:content/translations/en.DM00068118.pdf) which shows the table I drew. 

I don't have F0 develop board, so I can not set it support at default.

kevinong0108 replied:
This commit https://github.com/ARMmbed/mbed-os/commit/ba25cd4e1f83e71f629ca4a67b4580abb5900b67 modify F3 to support 7, 8, 16 bits.

jeromecoutant commented on targets/TARGET_STM/mbed_crc_api.c:
```diff
@@ -7,50 +7,103 @@
+	}
 }
 
 void hal_crc_compute_partial(const uint8_t *data, const size_t size)
```
I think issue with F1/L1/F2/F4 series, is this function ?

These families expect some u32 data, but they got only u8 data (only u32 cast)

kevinong0108 replied:
```
/**
  * @brief  Computes the 32-bit CRC of 32-bit data buffer using combination
  *         of the previous CRC value and the new one.
  * @param  hcrc pointer to a CRC_HandleTypeDef structure that contains
  *         the configuration information for CRC
  * @param  pBuffer pointer to the buffer containing the data to be computed
  * @param  BufferLength length of the buffer to be computed
  * @retval 32-bit CRC
  */
uint32_t HAL_CRC_Accumulate(CRC_HandleTypeDef *hcrc, uint32_t pBuffer[], uint32_t BufferLength)
```
The partial of `stm32f4xx_hal_crc.c` as above shows that use 4 bytes alignment only...
and that is very danger when data pointer is not 4 bytes address alignment with optimize option use -O2.

kevinong0108 replied:
I decide set F1/F2/F4/L1 to not supported, more detail please see https://github.com/ARMmbed/mbed-os/pull/9347/commits/ba25cd4e1f83e71f629ca4a67b4580abb5900b67

kevinong0108 committed:
stm32: fix F1/F2/F4/L1 compile fail after enable CRC

Those series have not CRC_POLYLENGTH_32B definition

kevinong0108 committed:
stm32: F1/F2/F4/L1 set to not supported HW CRC

Currently, mbed supported input data format fix on bytes, so those
devices are not supported at default.

0xc0170 commented:
CI started

0xc0170 commented:
CI job was aborted, 5.11.2 RC candidate is in the CI. We will restart it as soon as 5.11.2 RC is ready

cmonr commented:
CI started

cmonr commented:
@kevinong0108 Please take a look at the build failures.

kevinong0108 commented:
@cmonr OK, got it.

I can't open detail link of "jenkins-ci/build-GCC_ARM — Error", and I try to compile on my own environment is OK. @jeromecoutant has tested before too.

cmonr commented:
@kevinong0108 The `jenkins-ci/*` jobs are run on an internal CI.

The logs are updloaded to AWS, and provided in the `build artifacts` link in the build results comment: https://github.com/ARMmbed/mbed-os/pull/9347#issuecomment-454963620

kevinong0108 commented:
@cmonr oh~~ thanks!!! I will fix it asap.

kevinong0108 committed:
stm32: fix some F0 target compile error

kevinong0108 committed:
stm32: fix typo

kevinong0108 committed:
stm32: fix indent style

0xc0170 commented:
Travis failure is related to the master bug - the fix is in the CI now, should be resolved soon (rebase will be needed ,we can do it for you)

cmonr commented:
**NOTE:** This PR has now been rebased.

If this was made in error, feel free to force-push your local branch to restore the PR.

0xc0170 commented:
CI started

alekla01 commented:
restarted ci: greentea-test
